test/videojs-contrib-hls.test.js

```diff
@@ -1077,24 +1077,73 @@ QUnit.test('playlist 404 should blacklist media', function(assert) {
                            'media.m3u8\n' +
                            '#EXT-X-STREAM-INF:BANDWIDTH=1\n' +
                            'media1.m3u8\n');
-
   assert.equal(typeof this.player.tech_.hls.playlists.media_,
               'undefined',
               'no media is initially set');
-
   // media
-  this.requests[1].respond(400);
-
+  this.requests[1].respond(404);
   url = this.requests[1].url.slice(this.requests[1].url.lastIndexOf('/') + 1);
   media = this.player.tech_.hls.playlists.master.playlists[url];
-
   assert.ok(media.excludeUntil > 0, 'original media blacklisted for some time');
   assert.equal(this.env.log.warn.calls, 1, 'warning logged for blacklist');
+  assert.equal(this.env.log.warn.args[0],
+              'Problem encountered with the current HLS playlist. Switching to another playlist.',
+              'log generic error message');
+
+  // request for the final available media
+  this.requests[2].respond(404);
+  url = this.requests[2].url.slice(this.requests[2].url.lastIndexOf('/') + 1);
+  media = this.player.tech_.hls.playlists.master.playlists[url];
+
+  // media wasn't blacklisted because it's final rendition
+  assert.ok(!media.excludeUntil, 'media not blacklisted after playlist 404');
+  assert.equal(this.env.log.warn.calls, 1, 'warning logged for blacklist');
+  assert.equal(this.env.log.warn.args[1],
+              'Problem encountered with the current HLS playlist. Trying again since it is the final playlist.',
+              'log specific error message for final playlist');
+
+  this.clock.tick(2 * 1000);
+  // no more request was made since it hasn't been half the segment duration
+  assert.strictEqual(3, this.requests.length, 'no more request was made');
+
+  this.clock.tick(3 * 1000);
+  // continue loading the final remaining playlist after it wasn't blacklisted
+  // when half the segment duaration passed
+  assert.strictEqual(4, this.requests.length, 'one more request was made');
+  assert.strictEqual(this.requests[3].url,
+                     absoluteUrl('manifest/media1.m3u8'),
+                     'media playlist requested');
 
   // verify stats
   assert.equal(this.player.tech_.hls.stats.bandwidth, 1e10, 'bandwidth set above');
 });
 
+QUnit.test('never blacklist the playlist if it is the only playlist', function(assert) {
```

Comment:
I think we need one more test for the behavior of a playlist error (404 or 500) on the first request for that playlist (that it should fail fast when there is only that one media playlist and no master). It looks like this was changed such that it gets a response first and is never blacklisted, which is a good test to have, but we should have the other case too.